obp/ope/estimators_tuning.py

```diff
@@ -0,0 +1,956 @@
+
+        action: array-like, shape (n_rounds,)
+            Action sampled by a behavior policy in each round of the logged bandit feedback, i.e., :math:`a_t`.
+
+        pscore: array-like, shape (n_rounds,)
+            Action choice probabilities by a behavior policy (propensity scores), i.e., :math:`\\pi_b(a_t|x_t)`.
+
+        action_dist: array-like, shape (n_rounds, n_actions, len_list)
+            Action choice probabilities
+            by the evaluation policy (can be deterministic), i.e., :math:`\\pi_e(a_t|x_t)`.
+
+        position: array-like, shape (n_rounds,), default=None
+            Positions of each round in the given logged bandit feedback.
+
+        alpha: float, default=0.05
+            P-value.
+
+        n_bootstrap_samples: int, default=10000
+            Number of resampling performed in the bootstrap procedure.
+
+        random_state: int, default=None
+            Controls the random seed in bootstrap sampling.
+
+        Returns
+        ----------
+        estimated_confidence_interval: Dict[str, float]
+            Dictionary storing the estimated mean and upper-lower confidence bounds.
+
+        """
+        if not isinstance(reward, np.ndarray):
+            raise ValueError("reward must be ndarray")
+        if not isinstance(action, np.ndarray):
+            raise ValueError("action must be ndarray")
+        if not isinstance(pscore, np.ndarray):
+            raise ValueError("pscore must be ndarray")
+
+        check_ope_inputs(
+            action_dist=action_dist,
+            position=position,
+            action=action,
+            reward=reward,
+            pscore=pscore,
+        )
+        if position is None:
+            position = np.zeros(action_dist.shape[0], dtype=int)
+
+        # tune the clipping hyperparameter
+        self.estimated_mse_upper_bound_dict = dict()
+        for lambda_ in self.lambdas:
+            estimated_mse_upper_bound = InverseProbabilityWeighting(
+                lambda_=lambda_
+            )._estimate_mse_upper_bound(
+                reward=reward,
+                action=action,
+                position=position,
+                pscore=pscore,
+                action_dist=action_dist,
+            )
+            self.estimated_mse_upper_bound_dict[lambda_] = estimated_mse_upper_bound
+        self.best_lambda_ = min(
+            self.estimated_mse_upper_bound_dict.items(), key=lambda x: x[1]
+        )[0]
+
+        estimated_round_rewards = InverseProbabilityWeighting(
+            lambda_=self.best_lambda_
+        )._estimate_round_rewards(
+            reward=reward,
+            action=action,
+            position=position,
+            pscore=pscore,
+            action_dist=action_dist,
+        )
+        return estimate_confidence_interval_by_bootstrap(
+            samples=estimated_round_rewards,
+            alpha=alpha,
+            n_bootstrap_samples=n_bootstrap_samples,
+            random_state=random_state,
+        )
```

Comment:
(Question)
Is there a reason why you don't use `InverseProbabilityWeighting().estimate_interval()`?

Reply:
Fixed. Thanks!